docs/proposals/20210716-async-azure-resource-creation-deletion.md

```diff
@@ -0,0 +1,320 @@
+    RouteTablesReadyCondition clusterv1.ConditionType = "RouteTablesReady"
+    // PublicIPsReadyCondition means the public IPs exist and are ready to be used.
+    PublicIPsReadyCondition clusterv1.ConditionType = "PublicIPsReady"
+    // NATGatewaysReadyCondition means the NAT gateways exist and are ready to be used.
+    NATGatewaysReadyCondition clusterv1.ConditionType = "NATGatewaysReady"
+    // SubnetsReadyCondition means the subnets exist and are ready to be used.
+    SubnetsReadyCondition clusterv1.ConditionType = "SubnetsReady"
+    // LoadBalancersReadyCondition means the load balancers exist and are ready to be used.
+    LoadBalancersReadyCondition clusterv1.ConditionType = "LoadBalancersReady"
+    // PrivateDNSReadyCondition means the private DNS exists and is ready to be used.
+    PrivateDNSReadyCondition clusterv1.ConditionType = "PrivateDNSReady"
+    // BastionHostReadyCondition means the bastion host exists and is ready to be used.
+    BastionHostReadyCondition clusterv1.ConditionType = "BastionHostReady"
+    // InboundNATRulesReadyCondition means the inbound NAT rules exist and are ready to be used.
+    InboundNATRulesReadyCondition clusterv1.ConditionType = "InboundNATRulesReady"
+    // AvailabilitySetReadyCondition means the availability set exists and is ready to be used.
+    AvailabilitySetReadyCondition clusterv1.ConditionType = "AvailabilitySetReady"
+    // RoleAssignmentReadyCondition means the role assignment exists and is ready to be used.
+    RoleAssignmentReadyCondition clusterv1.ConditionType = "RoleAssignmentReady"
+
+    // CreatingReason means the resource is being created.
+    CreatingReason = "Creating"
+    // FailedReason means the resource failed to be created.
+    FailedReason = "Failed"
+    // DeletingReason means the resource is being deleted.
+    DeletingReason = "Deleting"
+    // DeletedReason means the resource was deleted.
+    DeletedReason = "Deleted"
+    // DeletionFailedReason means the resource failed to be deleted.
+    DeletionFailedReason = "DeletionFailed"
+)
+```
+
+### <a name='OpenQuestions'></a>Open Questions
+
+#### <a name='Whatshouldthetimeoutdurationsbe'></a>1. What should the timeout durations be?
+
+The specific numbers are not set in stone, and should be revised after doing some performance testing with different values and calculating the P99 expected completion time of operations that are not long-running.
+
+The other question is whether we should have the same timeout value for all operations (the 5s) or curate per operation. For simplicity, the proposal is to start with a single value. Later on, we might want to optimize by calculating a dynamic timeout value for each operation based on heuristics. That would be better than statically defining artificial timeout durations for each operation which might vary over time and might not be the same depending on region, subscription, etc.
```

Comment:
The reason to define a common timeout is that this timeout in practice is really just the timeout for getting a response from the Azure API. Is that correct?

If so, we want to make sure there are go Azure API GET calls that don't properly paginate when returning a set (or, probably more importantly, if the underlying resources themselves don't paginate their LIST responses). So long as Azure APIs return paginated sets generally, we should be able to make practical assumptions (e.g., 5 seconds) about how long before we want to timeout the API response. If certain APIs, however, don't paginate responses for LIST requests, then we could expect that the response time could scale with the number of items in the response.

Reply:
Table this comment